### PR TITLE
Scalar multiplication/quotient (implicit broadcasting)

### PR DESCRIPTION
**Pull request by pmichel31415**
This partially addresses #355 and #440 

Now,

    cmult(x,y);

Also works if either `x` or `y` are (optionally batched) scalars

Same for `cdiv(x,y)` if `y` is scalar

Should also transfer to the python API.

Basically this creates two nodes `ScalarMultiply` and `ScalarQuotient` and the code of expressions `cmult` and `cdiv` checks the dimensions and adds the appropriate node to the computation graph.

### CODE REVIEW

File: dynet/expr.cc [hunk 1]
```diff
@@ -47,8 +47,8 @@ Expression operator-(real x, const Expression& y) { return Expression(y.pg, y.pg
 Expression operator-(const Expression& x, real y) { return -(y - x); }
 Expression operator*(const Expression& x, const Expression& y) { return Expression(x.pg, x.pg->add_function<MatrixMultiply>({x.i, y.i})); }
 Expression operator*(const Expression& x, float y) { return Expression(x.pg, x.pg->add_function<ConstScalarMultiply>({x.i}, y)); }
-Expression cmult(const Expression& x, const Expression& y) {return Expression(x.pg, x.pg->add_function<CwiseMultiply>({x.i, y.i}));}
-Expression cdiv(const Expression& x, const Expression& y) { return Expression(x.pg, x.pg->add_function<CwiseQuotient>({x.i, y.i})); }
+Expression cmult(const Expression& x, const Expression& y) { if (x.dim().batch_size() == 1) return Expression(x.pg, x.pg->add_function<ScalarMultiply>({x.i, y.i})); else if(y.dim().batch_size()==1)return Expression(x.pg, x.pg->add_function<ScalarMultiply>({y.i, x.i})); else return Expression(x.pg, x.pg->add_function<CwiseMultiply>({x.i, y.i}));}
```
Comment (neubig) [inline dynet/expr.cc]: Can we make this multiple lines? Same with the following.

**Comment by pmichel31415** [comment]
OK I fixed both, should be ready to merge when the tests finish.

I agree that a centralized auto formatter would be nice. I personally use the SublimeAStyleFormatter plugin for sublime text, I think that it's based on [AStyle](http://astyle.sourceforge.net/), might be worth considering

**Comment by michaelcapizzi** [comment]
Thanks for implementing this, @pmichel31415.  I'm actually trying to "embedding-level" dropout as you described in the attached paper in #355 .

But I have a question about backpropogation.

If I "zero-out" an embedding after `batch_lookup()` using `cmult()`, will the error *still* get backpropogated to the `lookup` parameters for that embedding *even though* it was `dropped-out` in that instance?

**Comment by pmichel31415** [comment]
The gradient will be 0 since multiplication by 0 will also be performed during the backward computation

ie if `ê = m * e` where `e` is the embedding and `m` is the dropout mask, `dL/de = m * dL/dê` where `L` is your objective.

**Comment by michaelcapizzi** [comment]
Ah!  Makes sense.  Perfect.

Thank you @pmichel31415 for your quick response.

**Pull request merged**